File: Data Structures & Algorithms/balanced-binary-tree/submission-4.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isBalanced(root) {
        function dfs(node) {
            if (!node) return 0;

            const leftHeight = dfs(node.left);
            if (leftHeight === -1) return -1;

            const rightHeight = dfs(node.right);
            if (rightHeight === -1) return -1;

            if (Math.abs(leftHeight - rightHeight) > 1) return -1;

            return 1 + Math.max(leftHeight, rightHeight); 
        }

        return dfs(root) !== -1;

    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-4.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and not subRoot:
            return True

    
        def isSameTree(p, q):
            if not p and not q:
                return True

            if not p or not q:
                return False

            if p.val != q.val:
                return False

            return isSameTree(p.right, q.right) and isSameTree(p.left, q.left)

        def dfs(root, subRoot):
            if not root:
                return False

            if root.val == subRoot.val and isSameTree(root, subRoot):
                return True

            return dfs(root.right, subRoot) or dfs(root.left, subRoot)

        return dfs(root, subRoot)